### ISSUE TITLE
Unresolved reference error for snakemake specific python methods and objects

### ISSUE THREAD

**Issue by dievsky** [JetBrains-Research/snakecharm#1]
Built-in Snakemake objects and methods `config, directory, rules, temp, expand` are not recognized as such.

**Comment by iromeo** [comment]
E.g.: 

`rules`

```
rule bam_to_bdg:
    input:
        bam="bams/{sample}.bam",
        cs=rules.download_chrom_sizes.output
    output: temp("bws/{sample}.bdg")
    shell: 'bedtools genomecov -ibam {input.bam} -bg -g {input.cs} > {output}'
```

----
`config`
```
```

----
`directory`
```
```
----
`temp`
```
```
----
`expand`
```
```
----

**Comment by iromeo** [comment]
E.g. `shell` is given from `from snakemake.shell import shell`

**Comment by dievsky** [comment]
Also `protected`:
```Python
rule x:
    output: protected("a.txt")
```

**Comment by iromeo** [comment]
`ancient`
https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html#ignoring-timestamps

```
rule NAME:
    input:
        ancient("path/to/inputfile")
    output:
        "path/to/outputfile"
    shell:
        "somecommand {input} {output}"
```

**Comment by flying-sheep** [comment]
Also `input`/`output`:

```snakemake
rule NAME:
    input:
        "path/to/inputfile"
    output:
        "path/to/outputfile"
    run:
        some_func(input[0], output[0])
```

`input` is resolved as the builtin `input()` function, output not at all.

**Comment by iromeo** [comment]
@flying-sheep also `params`, `threads`, `wildcards` objects

**Comment by flying-sheep** [comment]
True. `wildcards` is a `snakemake.io.Wildcards`, `threads` is an `int` I supposed, and `params`?

**Comment by iromeo** [comment]
@flying-sheep I think for `wildcards`, `params` we need to hardcode resolve/autocompletion to rule wildcard strings or `params` section named args. The true runtime type here isn't very significant.

**Comment by iromeo** [comment]
Done
![image](https://user-images.githubusercontent.com/72933/60993668-96438380-a357-11e9-99d1-2b05c2b65290.png)

**Comment by lpla** [comment]
I don't see this fixed in latest release 0.4.35.2018.3. I am still having this `Undefined reference` error with temp, output, config, expand, shell, wildcards and maybe more reserved words.

My Snakefile is here: https://github.com/bitextor/bitextor/blob/master/snakemake/Snakefile

**Comment by iromeo** [comment]
@lpla we decided not to hardcode snakemake specific method names in the plugin. Instead we collect them from snakamake package and theoretically will automatically support new versions of snakemake if authors add/remove some methods, global variables, etc. That is why it is required to setup python interpreter with snakemake package in ide settings.
